File: apps/api/src/services/resume.service.ts
// Import the internal module directly, not the package root — pdf-parse's
// index.js runs a debug demo block at import time whenever `module.parent`
// is falsy (true under tsx's module loader), which tries to read a
// nonexistent fixture PDF and crashes the whole process on startup.
import pdfParse from "pdf-parse/lib/pdf-parse.js";
import type { HydratedDocument } from "mongoose";
import type { Resume } from "@believe-ai/shared";
import type { ResumeDocument } from "@believe-ai/server";
import { resumeRepository } from "../repositories/resume.repository.js";
import { embedTexts } from "./aiServiceClient.js";
import { chunkText } from "../utils/chunkText.js";
import { NotFoundError, ValidationError } from "../errors/AppError.js";
import { logger } from "../config/logger.js";

const MAX_RESUME_BYTES = 4 * 1024 * 1024;

function toDto(doc: HydratedDocument<ResumeDocument>): Resume {
  return {
    id: doc._id.toString(),
    userId: doc.userId.toString(),
    fileName: doc.fileName,
    mimeType: doc.mimeType,
    sizeBytes: doc.sizeBytes,
    content: doc.content,
    chunks: doc.chunks.map((c) => ({ text: c.text, vector: c.vector ?? null })),
    embeddingReady: doc.chunks.length > 0 && doc.chunks.every((c) => c.vector != null),
    createdAt: doc.createdAt.toISOString(),
    updatedAt: doc.updatedAt.toISOString(),
  };
}

export const resumeService = {
  toDto,

  async getByUserId(userId: string): Promise<Resume> {
    const resume = await resumeRepository.findByUserId(userId);
    if (!resume) throw new NotFoundError("No resume uploaded yet");
    return toDto(resume);
  },

  /**
   * Upload replaces any existing resume wholesale (stale chunks/vectors from
   * a previous file can never survive). Parsing and chunking happen here,
   * not in apps/ai-service, so a resume can be stored and browsed even if
   * every AI provider is down — only the embedding step depends on AI, and
   * a partially-embedded resume degrades gracefully (embeddingReady: false)
   * rather than failing the whole upload.
   */
  async upload(userId: string, file: { originalname: string; mimetype: string; size: number; buffer: Buffer }, bearerToken: string): Promise<Resume> {
    if (file.mimetype !== "application/pdf") {
      throw new ValidationError("Only PDF resumes are supported right now");
    }
    if (file.size > MAX_RESUME_BYTES) {
      throw new ValidationError("Resume must be under 4 MB");
    }

    const parsed = await pdfParse(file.buffer).catch(() => {
      throw new ValidationError("Could not read this PDF — it may be corrupted or scanned as images");
    });
    const content = parsed.text.trim();
    if (!content) {
      throw new ValidationError("No extractable text found in this PDF");
    }

    const chunks = chunkText(content).map((text) => ({ text, vector: null as number[] | null }));

    await resumeRepository.upsertFile(userId, file.originalname, file.mimetype, file.buffer);
    const resume = await resumeRepository.upsert({
      userId,
      fileName: file.originalname,
      mimeType: file.mimetype,
      sizeBytes: file.size,
      content,
      chunks,
    });

    // Embedding failure shouldn't fail the upload — the resume is stored and
    // usable (e.g. for direct display) even before it's RAG-ready. The
    // caller can re-trigger embedding by re-uploading.
    try {
      const { vectors } = await embedTexts(
        bearerToken,
        chunks.map((c) => c.text),
      );
      const embedded = await resumeRepository.setChunkVectors(userId, vectors);
      return toDto(embedded ?? resume);
    } catch (err) {
      logger.warn({ err, userId }, "Resume embedding failed, resume stored without vectors");
      return toDto(resume);
    }
  },

  async delete(userId: string): Promise<void> {
    const [resume] = await resumeRepository.delete(userId);
    if (!resume) throw new NotFoundError("No resume uploaded yet");
  },
};

File: apps/backend/services/resume_service.py
"""The caller's own resumes — mirrors apps/api's resume.service.ts, extended
for many-resumes-per-user. Parsing and chunking happen here, not behind a
separate AI call, so a resume can be stored and browsed even if every AI
provider is down — only the embedding step depends on AI, and a
partially-embedded resume degrades gracefully (embeddingReady: false) rather
than failing the whole upload."""

from __future__ import annotations

import io

from bson import ObjectId
from pypdf import PdfReader
from pypdf.errors import PdfReadError

from agents.embedding_agent import embed_texts
from core.config import Settings
from core.errors import NotFoundError, ValidationError
from core.logging import get_logger
from models.resume import Resume, ResumeChunk
from repositories import resume_repository
from schemas.ai import EmbedTextsRequest
from schemas.resume import ResumeChunkDto, ResumeDto
from utils.chunk_text import chunk_text

logger = get_logger(__name__)

_MAX_RESUME_BYTES = 4 * 1024 * 1024


def _to_dto(doc: Resume) -> ResumeDto:
    assert doc.id is not None
    return ResumeDto(
        id=str(doc.id),
        userId=str(doc.userId),
        fileName=doc.fileName,
        mimeType=doc.mimeType,
        sizeBytes=doc.sizeBytes,
        content=doc.content,
        chunks=[ResumeChunkDto(text=c.text, vector=c.vector) for c in doc.chunks],
        embeddingReady=len(doc.chunks) > 0 and all(c.vector is not None for c in doc.chunks),
        isPrimary=doc.isPrimary,
        targetRole=doc.targetRole,
        createdAt=doc.createdAt.isoformat(),
        updatedAt=doc.updatedAt.isoformat(),
    )


def _extract_pdf_text(data: bytes) -> str:
    try:
        reader = PdfReader(io.BytesIO(data))
        text = "\n".join(page.extract_text() or "" for page in reader.pages)
    except PdfReadError as err:
        raise ValidationError("Could not read this PDF — it may be corrupted or scanned as images") from err
    return text.strip()


async def list_for_user(user_id: ObjectId) -> list[ResumeDto]:
    resumes = await resume_repository.list_for_user(user_id)
    return [_to_dto(doc) for doc in resumes]


async def get_by_id(user_id: ObjectId, resume_id: ObjectId) -> ResumeDto:
    resume = await resume_repository.find_by_id(user_id, resume_id)
    if not resume:
        raise NotFoundError("Resume not found")
    return _to_dto(resume)


async def upload(settings: Settings, user_id: ObjectId, file_name: str, mime_type: str, data: bytes, target_role: str | None) -> ResumeDto:
    if mime_type != "application/pdf":
        raise ValidationError("Only PDF resumes are supported right now")
    if len(data) > _MAX_RESUME_BYTES:
        raise ValidationError("Resume must be under 4 MB")

    content = _extract_pdf_text(data)
    if not content:
        raise ValidationError("No extractable text found in this PDF")

    chunks = [ResumeChunk(text=text, vector=None) for text in chunk_text(content)]

    resume = await resume_repository.create(user_id, file_name, mime_type, len(data), content, chunks, target_role)
    assert resume.id is not None
    await resume_repository.create_file(resume.id, user_id, file_name, mime_type, data)

    # Embedding failure shouldn't fail the upload — the resume is stored and
    # usable (e.g. for direct display) even before it's RAG-ready. The
    # caller can re-trigger embedding by deleting and re-uploading.
    try:
        result = await embed_texts(settings, EmbedTextsRequest(texts=[c.text for c in chunks]))
        embedded = await resume_repository.set_chunk_vectors(user_id, resume.id, list(result.vectors))
        return _to_dto(embedded or resume)
    except Exception as err:  # noqa: BLE001 — see docstring: a failed embed must never fail the upload
        logger.warning("Resume embedding failed for user %s, resume stored without vectors: %s", user_id, err)
        return _to_dto(resume)


async def set_primary(user_id: ObjectId, resume_id: ObjectId) -> ResumeDto:
    resume = await resume_repository.set_primary(user_id, resume_id)
    if not resume:
        raise NotFoundError("Resume not found")
    return _to_dto(resume)


async def update(user_id: ObjectId, resume_id: ObjectId, target_role: str | None) -> ResumeDto:
    resume = await resume_repository.update(user_id, resume_id, target_role)
    if not resume:
        raise NotFoundError("Resume not found")
    return _to_dto(resume)


async def delete(user_id: ObjectId, resume_id: ObjectId) -> None:
    deleted = await resume_repository.delete_by_id(user_id, resume_id)
    if not deleted:
        raise NotFoundError("Resume not found")
